[iOS] Use unique temp directory for each WebKit process type

pvollan opened the pull request:
#### b21c5898afd2dd1fe158c129f1613d6925e2fdc0
<pre>
[iOS] Use unique temp directory for each WebKit process type
<a href="https://bugs.webkit.org/show_bug.cgi?id=291083">https://bugs.webkit.org/show_bug.cgi?id=291083</a>
<a href="https://rdar.apple.com/148592903">rdar://148592903</a>

Reviewed by Chris Dumez.

Use a unique temp folder for each WebKit process type in the parent process&apos; data container.

* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/WebKitServiceNames.h: Added.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):

Canonical link: <a href="https://commits.webkit.org/293291@main">https://commits.webkit.org/293291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d8a24a0eddf57159c4996d251a0a687f6ea7648

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18021 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8252 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48918 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26473 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74898 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32070 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13893 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88862 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55257 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13673 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48360 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83627 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6904 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105884 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25478 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83877 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85063 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83354 "Found 1 new API test failure: /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27994 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5667 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19131 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15962 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25437 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25255 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28575 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26830 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -183,19 +184,21 @@ void XPCServiceEventHandler(xpc_connection_t peer)
                 setenv(key, xpc_string_get_string_ptr(value), 1);  // NOLINT
                 return true;
             });
+            if (!_set_user_dir_suffix(serviceName.ascii().data()))
```
I think `latin1()` is actually better usually as this is what the string is stored as internally and doesn't require any transformation.

cdumez commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -190,7 +190,8 @@ GPUProcessProxy::GPUProcessProxy()
 
     if (!containerTemporaryDirectory.isEmpty()) {
-        if (auto handle = SandboxExtension::createHandleWithoutResolvingPath(containerTemporaryDirectory, SandboxExtension::Type::ReadWrite))
+        auto tempDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(makeString(containerTemporaryDirectory, "/com.apple.WebKit.GPU"_s));
```
Is it really OK to hard-code the `/` here? Shouldn't we use the file system API to append the directory?

pvollan replied:
Good point, will fix!

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -890,8 +890,11 @@ WebProcessDataStoreParameters WebProcessPool::webProcessDataStoreParameters(WebP
-        containerTemporaryDirectoryExtensionHandle = SandboxExtension::createHandleWithoutResolvingPath(directory, SandboxExtension::Type::ReadWrite);
+    if (auto directory = websiteDataStore.resolvedContainerTemporaryDirectory(); !directory.isEmpty()) {
+        if (m_cachedWebContentTempDirectory.isEmpty())
+            m_cachedWebContentTempDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(makeString(directory, "/com.apple.WebKit.WebContent"_s));
```
ditto.

szewai commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -183,19 +184,21 @@ void XPCServiceEventHandler(xpc_connection_t peer)
                 setenv(key, xpc_string_get_string_ptr(value), 1);  // NOLINT
                 return true;
             });
+            if (!_set_user_dir_suffix(serviceName.latin1().data()))
```
Can we log for both succeed and fail cases? Just to make it easy to notice change on temp directory is from log

pvollan replied:
Will do! Thanks for reviewing!

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2234,7 +2234,7 @@ WebsiteDataStoreParameters WebsiteDataStore::parameters()
         parameters.parentBundleDirectoryExtensionHandle = WTFMove(parentBundleDirectoryExtensionHandle);
 
-        if (auto handleAndFilePath = SandboxExtension::createHandleForTemporaryFile(emptyString(), SandboxExtension::Type::ReadWrite))
+        if (auto handleAndFilePath = SandboxExtension::createHandleForTemporaryFile("com.apple.WebKit.Networking"_s, SandboxExtension::Type::ReadWrite))
```
Is there a way to get the string from service-name instead of hard coding it? I think TestRunner might use "com.apple.WebKit.Networking.Development" on iOS simulator.
At least we probably want some test or assert to ensure UI process grants extension to a path that child process expects.

cdumez reviewed (approved):
LGTM

pvollan committed:
[iOS] Use unique temp directory for each WebKit process type
https://bugs.webkit.org/show_bug.cgi?id=291083
rdar://148592903

Reviewed by Chris Dumez.

Use a unique temp folder for each WebKit process type in the parent process' data container.

* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/WebKitServiceNames.h: Added.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::parameters):

Canonical link: https://commits.webkit.org/293291@main
